infrastructure/src/main/java/org/corfudb/infrastructure/ServerHandshakeHandler.java

```diff
@@ -0,0 +1,213 @@
+            if (this.handshakeFailed.get()) {
+                return;
+            }
+
+            if (this.handshakeComplete.get()) {
+                // If handshake completed successfully, but still a message came through this handler,
+                // send on to the next handler in order to avoid message loss.
+                super.channelRead(ctx, m);
+                return;
+            }
+
+            CorfuPayloadMsg<HandshakeMsg> handshake;
+
+            try {
+                handshake = (CorfuPayloadMsg<HandshakeMsg>) m;
+            } catch (ClassCastException e) {
+                log.warn("Non-handshake message received by handshake handler. Send upstream only " +
+                        "if handshake succeeded.");
+                if (this.handshakeComplete.get()) {
+                    // Only send upstream if handshake is complete.
+                    super.channelRead(ctx, m);
+                }
+                return;
+            }
+
+            UUID clientId = handshake.getPayload().getClientId();
+            UUID serverId = handshake.getPayload().getServerId();
+
+            // Validate handshake
+            if (!serverId.equals(this.nodeId)) {
+                log.error("Invalid handshake: this is " + this.nodeId +
+                        " and client is trying to connect to " + serverId);
+                this.fireHandshakeFailed(ctx);
+                return;
+            }
+
+            // Store clientID as a channel attribute.
+            AttributeKey<UUID> clientIdAttrKey = AttributeKey.valueOf("ClientID");
+            ctx.channel().attr(clientIdAttrKey).set(clientId);
+            log.info("Handshake validated by Server.");
+            log.debug("Sending handshake response: Node Id: " + this.nodeId +
+                    " Corfu Version: " + this.corfuVersion);
+
+            CorfuMsg handshakeResponse = CorfuMsgType.HANDSHAKE_RESPONSE
+                    .payloadMsg(new HandshakeResponse(this.nodeId, this.corfuVersion));
+            ctx.writeAndFlush(handshakeResponse);
+
+            // Flush messages in queue
+            log.debug("There are {" + this.messages.size() + "} messages in queue to " +
+                    "be flushed.");
+            for (CorfuMsg message : this.messages) {
+                ctx.writeAndFlush(message);
+            }
+
+            // Remove this handler from the pipeline; handshake is completed.
+            log.info("Removing handshake handler from pipeline.");
+            ctx.pipeline().remove(this);
+            this.fireHandshakeSucceeded();
+        }
+    }
+
+
+    @Override
+    public void channelActive(final ChannelHandlerContext ctx) throws Exception {
+        this.group.add(ctx.channel());
+        log.info("Incoming connection established from: " +
+                ctx.channel().remoteAddress());
+
+        // Start timeout checker. Wait for handshake to complete, otherwise close connection.
+        new Thread() {
```

Comment:
Hm. I think we want to avoid creating a new thread here. I think we have two options here, we can schedule on the executor for the context (```ctx.executor()```), or we can use the ```ReadTimeoutHandler``` and uninstall it from the pipeline as well (I think you can uninstall by name). I'm sort of leaning towards the ReadTimeoutHandler.

Reply:
I wasn't aware of the ReadTimeoutHandler, it is really useful. I also lean towards this approach. What I am doing now is adding the ReadTimeoutHandler upon handshake initiation, and removing it on ChannelRead.. This also removes the latch and the mutex